node/block_exchange.go

```diff
@@ -0,0 +1,226 @@
+	if trustedBlock != nil {
+		if err := bExService.initBlockStoreAndStartSyncer(ctx, trustedBlock); err != nil {
+			bExService.logger.Error("failed to initialize the headerstore and start syncer", "error", err)
+		}
+	}
+}
+
+func (bExService *BlockExchangeService) writeToBlockStoreAndBroadcast(ctx context.Context, signedHeader *types.SignedHeader) {
+	// For genesis header initialize the store and start the syncer
+	if signedHeader.Height() == bExService.genesis.InitialHeight {
+		if err := bExService.blockStore.Init(ctx, signedHeader); err != nil {
+			bExService.logger.Error("failed to initialize header store", "error", err)
+		}
+
+		if err := bExService.syncer.Start(bExService.ctx); err != nil {
+			bExService.logger.Error("failed to start syncer after initializing header store", "error", err)
+		}
+	}
+
+	// Broadcast for subscribers
+	if err := bExService.sub.Broadcast(ctx, signedHeader); err != nil {
+		bExService.logger.Error("failed to broadcast block header", "error", err)
+	}
+}
+
+// OnStart is a part of Service interface.
+func (bExService *BlockExchangeService) Start() error {
+	var err error
+	// have to do the initializations here to utilize the p2p node which is created on start
+	ps := bExService.p2p.PubSub()
+	bExService.sub = goheaderp2p.NewSubscriber[*types.SignedHeader](ps, pubsub.DefaultMsgIdFn, bExService.genesis.ChainID)
+	if err = bExService.sub.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting subscriber: %w", err)
+	}
+	if _, err := bExService.sub.Subscribe(); err != nil {
+		return fmt.Errorf("error while subscribing: %w", err)
+	}
+
+	if err = bExService.blockStore.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting header store: %w", err)
+	}
+
+	_, _, network := bExService.p2p.Info()
+	if bExService.p2pServer, err = newP2PServer(bExService.p2p.Host(), bExService.blockStore, network); err != nil {
+		return err
+	}
+	if err = bExService.p2pServer.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting p2p server: %w", err)
+	}
+
+	peerIDs := bExService.p2p.PeerIDs()
+	if bExService.ex, err = newP2PExchange(bExService.p2p.Host(), peerIDs, network, bExService.genesis.ChainID, bExService.p2p.ConnectionGater()); err != nil {
+		return err
+	}
+	if err = bExService.ex.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting exchange: %w", err)
+	}
+
+	if bExService.syncer, err = newSyncer(bExService.ex, bExService.blockStore, bExService.sub, goheadersync.WithBlockTime(bExService.conf.BlockTime)); err != nil {
+		return err
+	}
+
+	// Check if the headerstore is not initialized and try initializing
+	if bExService.blockStore.Height() > 0 {
+		if err := bExService.syncer.Start(bExService.ctx); err != nil {
+			return fmt.Errorf("error while starting the syncer: %w", err)
+		}
+		bExService.syncerStatus.started = true
```

Comment:
this is a perfect example of why the status should be set in the `Start` method. Here we are editing the field outside of a mutex lock.

Reply:
Talked about this with Ganesh, and he says that this field is never touched by another method while we do this. I think it's better practice to just make this a method.

Reply:
> he says that this field is never touched by another method while we do this. 

I don't think this should be how we think about it. This may be true today, but this could change in the future. We should write code that is future proof, which means protecting against potential future data races. If a struct has a mutex, there is no reason to not use imo. There is virtually zero performance impact and making this a habit will lead to more robust code.

Reply:
Agreed, which is why I made a method that uses the mutex :) cc @gupadhyaya for the above comment